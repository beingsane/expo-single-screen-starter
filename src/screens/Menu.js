import React from 'react';
import { Animated, ScrollView, StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { colors, device } from '../api/constants';

// component(s)
import Touch from '../components/Touch';

class Menu extends React.Component {
  constructor(props) {
    super(props);

    let positionStart = -device.width;
    if (props.direction === 'right') {
      positionStart = device.width;
    }

    this.state = {
      backgroundColor: new Animated.Value(0),
      position: new Animated.Value(positionStart),
      positionStart
    };

    this.handleClose = this.handleClose.bind(this);
  }

  componentDidMount() {
    // console.log('Menu::componentDidMount()');
    const { backgroundColor, position } = this.state;

    Animated.timing(backgroundColor, {
      delay: 200,
      duration: 400,
      toValue: 10
    }).start();

    Animated.timing(position, {
      toValue: 0,
      duration: 400
    }).start();
  }

  // componentWillUnmount() {
  // console.log('Menu::componentWillUnmount()');
  // }

  handleClose() {
    const { onClose } = this.props;
    const { backgroundColor, position, positionStart } = this.state;

    Animated.timing(backgroundColor, {
      toValue: 0,
      duration: 400
    }).start();

    Animated.timing(position, {
      toValue: positionStart,
      duration: 400
    }).start(() => {
      onClose();
    });
  }

  renderLayout() {
    const { direction } = this.props;
    const { backgroundColor } = this.state;

    let render = '';
    const color = backgroundColor.interpolate({
      inputRange: [0, 10],
      outputRange: [colors.white0, colors.white20]
    });

    if (direction === 'left') {
      render = (
        <Animated.View
          style={{
            backgroundColor: color,
            alignItems: 'stretch',
            flex: 1,
            flexDirection: 'row',
            justifyContent: 'flex-start',
            height: '100%',
            width: '100%'
          }}
        >
          <ScrollView
            style={{
              backgroundColor: colors.brandPrimary,
              height: '100%',
              paddingTop: device.iPhoneX ? 44 : 0,
              width: '80%'
            }}
          >
            <View
              style={{
                backgroundColor: colors.white,
                height: 75,
                marginBottom: 20,
                width: '100%'
              }}
            />
            <View
              style={{
                backgroundColor: colors.white,
                height: 75,
                marginBottom: 20,
                width: '100%'
              }}
            />
            <View
              style={{
                backgroundColor: colors.white,
                height: 75,
                marginBottom: 20,
                width: '100%'
              }}
            />
            <View
              style={{
                backgroundColor: colors.white,
                height: 75,
                marginBottom: 20,
                width: '100%'
              }}
            />
          </ScrollView>
          <Touch
            onPress={() => this.handleClose()}
            style={{
              height: '100%',
              width: '20%'
            }}
          />
        </Animated.View>
      );
    } else if (direction === 'right') {
      render = (
        <View
          style={{
            backgroundColor: colors.white50,
            flex: 1,
            flexDirection: 'row'
          }}
        >
          <Touch
            onPress={() => this.handleClose()}
            style={{
              height: '100%',
              width: '20%'
            }}
          />
          <ScrollView
            style={{
              backgroundColor: 'blue',
              height: '100%',
              width: '80%'
            }}
          >
            <View
              style={{
                backgroundColor: colors.white,
                height: 75,
                width: '100%'
              }}
            />
          </ScrollView>
        </View>
      );
    }

    return render;
  }

  render() {
    const { show } = this.props;
    const { position } = this.state;

    if (show === false) {
      return null;
    }

    return (
      <Animated.View style={[styles.container, { left: position }]}>
        {this.renderLayout()}
      </Animated.View>
    );
  }
}

// default props
Menu.defaultProps = {
  direction: 'left'
};

// type checking
Menu.propTypes = {
  // optional
  direction: PropTypes.string
};

const styles = StyleSheet.create({
  container: {
    height: '100%',
    position: 'absolute',
    width: '100%',
    zIndex: 100
  }
});

export default Menu;
