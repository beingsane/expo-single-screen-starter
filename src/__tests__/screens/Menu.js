import React from 'react';
import renderer from 'react-test-renderer';

import Menu from '../../screens/Menu';

it('Menu screen hidden', () => {
  jest.useFakeTimers();
  const onCloseEvent = jest.fn();
  onCloseEvent.mockReturnValue('Menu on close invoked');

  const tree = renderer.create(<Menu onClose={onCloseEvent} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Menu screen from left renders correctly', () => {
  jest.useFakeTimers();
  const onCloseEvent = jest.fn();
  onCloseEvent.mockReturnValue('Menu on close invoked');

  const tree = renderer.create(<Menu show onClose={onCloseEvent} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Menu screen from right renders correctly', () => {
  jest.useFakeTimers();
  const onCloseEvent = jest.fn();
  onCloseEvent.mockReturnValue('Menu on close invoked');

  const tree = renderer
    .create(<Menu direction="right" show onClose={onCloseEvent} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
