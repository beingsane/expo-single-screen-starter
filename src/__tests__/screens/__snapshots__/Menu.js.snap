// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu screen from left renders correctly 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "height": "100%",
      "left": -750,
      "position": "absolute",
      "width": "100%",
      "zIndex": 50,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "opacity": 1,
          "width": "20%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`Menu screen from right renders correctly 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "height": "100%",
      "left": 750,
      "position": "absolute",
      "width": "100%",
      "zIndex": 50,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "opacity": 1,
          "width": "20%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`Menu screen hidden 1`] = `null`;
